### PR TITLE
feature(load-and-stream): add nemesis

### PR DESCRIPTION
**Pull request by juliayakovlev**
Issue:
https://github.com/scylladb/scylla/commit/df3ef800c20c60d7929ffa600649793aa6c73064

Task:
https://trello.com/c/nwktmFqq/3478-dtest-7831-the-load-and-stream-feature-has-been-merged-this-allows-nodetool-refresh-to-split-loaded-sstables-into-owned-and-non

This feature extends the nodetool refresh to allow loading arbitrary sstables
that do not belong to a node into the cluster. It loads the sstables from disk
and calculates the owning nodes of the data and streams to the owners
automatically.

New nemesis runs refresh with 'load_and_stream' parameter and validate by the log that 
load_and_stream has been started and finished successfully.

## PR pre-checks (self review)

- [x] I followed [KISS principle](https://en.wikipedia.org/wiki/KISS_principle) and [best practices](https://docs.google.com/document/d/1jihgOKb5iGRlD8_HQ92O0JbLk1kASUoZT23i_MXFSKI)
- [x] I didn't leave commented-out/debugging code
- [x] I added the relevant `backport` labels
- [ ] New configuration option are added and documented (in `sdcm/sct_config.py`)
- [x] I have added tests to cover my changes (Infrastructure only - under `unit-test/` folder)
- [x] All new and existing unit tests passed (CI)
- [ ] I have updated the Readme/doc folder accordingly (if needed)

### CODE REVIEW

**Review by aleksbykov — DISMISSED** [review]
LGTM

File: data_dir/test_data_library/load_inventory.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+# Copyright (c) 2021 ScyllaDB
+"""
+Keep AWS S3 info about test sstables for refresh nemesis
+"""
+
+import copy
+from typing import NamedTuple, Any
+
+
+class TestDataInventory(NamedTuple):
```
[the remark this answers is not in the document]
Comment (bentsi) [reply data_dir/test_data_library/load_inventory.py:21]: those are constants hence those objects should be immutable. Namedtuple is immutable.

[the remark this answers is not in the document]
Comment (aleksbykov) [reply data_dir/test_data_library/load_inventory.py:21]: got it. thanks

**Comment by roydahan** [comment]
@juliayakovlev I', still reviewing... it's longer than I expected, but looks like you had fun :)

**Review by dkropachev — CHANGES_REQUESTED** [review]

File: data_dir/test_data_library/load_inventory.py [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+            keys_num=500000),
+        TestDataInventory(
+            sstable_url=URL_TEMPLATE % (LOAD_AND_STREAM_FOLDER, f'node4.{STANDARD1_FILE_NAME}'),
+            sstable_file=TEMP_FILE_TEMPLATE % f'node4.{STANDARD1_FILE_NAME}',
+            sstable_md5='e7ff37333c1e331acac58103e6e0406d',
+            keys_num=500000)
+     ]
+
+
+def get_load_test_data_inventory(column_number: int, big_sstable: bool, load_and_stream: bool) -> Any:
```
[the remark this answers is not in the document]
Comment (juliayakovlev) [reply data_dir/test_data_library/load_inventory.py:96]: done

File: data_dir/test_data_library/load_inventory.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+        # they rest columns will be filled to 'null'.
+        if load_and_stream:
+            return copy.deepcopy(MULTI_NODE_DATA)
+
+        if big_sstable:
+            return BIG_SSTABLE_MULTI_COLUMNS_DATA
+
+        return MULTI_COLUMNS_DATA
+
+    return None
```
[the remark this answers is not in the document]
Comment (juliayakovlev) [reply data_dir/test_data_library/load_inventory.py:115]: done

File: sdcm/utils/common.py [hunk 4]
```diff
@@ -2420,3 +2421,165 @@ def clear_out_all_exit_hooks():
...
+
+            if mod:
+                for _ in range(len(test_data)):
+                    node_to_load_to = self.get_random_item(nodes)
+                    map_files_to_node.append([self.get_random_item(test_data, pop=True), node_to_load_to])
+
+        return map_files_to_node
+
+    @staticmethod
+    def upload_sstables(node, test_data: namedtuple, keyspace_name: str = 'keyspace1'):
```
[the remark this answers is not in the document]
Comment (juliayakovlev) [reply sdcm/utils/common.py:2474]: done

File: sdcm/utils/common.py [hunk 5]
```diff
@@ -2420,3 +2421,165 @@ def clear_out_all_exit_hooks():
...
+        return len(re.findall(r"(\| C\d+)", result.stdout))
+
+    @staticmethod
+    def get_random_item(items: list, pop: bool = False) -> Any:
+        if pop:
+            return items.pop(random.randint(0, len(items) - 1))
+
+        return items[random.randint(0, len(items) - 1)]
+
+    def distribute_test_files_to_cluster_nodes(self, nodes, test_data: list) -> List:
```
[the remark this answers is not in the document]
Comment (juliayakovlev) [reply sdcm/utils/common.py:2445]: done

File: sdcm/utils/common.py [hunk 6]
```diff
@@ -2420,3 +2421,165 @@ def clear_out_all_exit_hooks():
...
+        else:
+            node.remoter.sudo(
+                f'tar xvfz {test_data.sstable_file} -C /var/lib/scylla/data/{keyspace_name}/{upload_dir}/upload/',
+                user='scylla')
+
+        # Scylla Enterprise 2019.1 doesn't support to load schema.cql and manifest.json, let's remove them
+        node.remoter.sudo(f'rm -f /var/lib/scylla/data/{keyspace_name}/{upload_dir}/upload/schema.cql')
+        node.remoter.sudo(f'rm -f /var/lib/scylla/data/{keyspace_name}/{upload_dir}/upload/manifest.json')
+
+    def run_load_and_stream(self, node, keyspace_name: str = 'keyspace1', table_name: str = 'standard1'):
```
[the remark this answers is not in the document]
Comment (juliayakovlev) [reply sdcm/utils/common.py:2495]: done

File: sdcm/utils/common.py [hunk 7]
```diff
@@ -2420,3 +2421,165 @@ def clear_out_all_exit_hooks():
...
+            for one_file in re.findall(r"(/var/.*?),", line, re.IGNORECASE):
+                # The file path have to include "upload" folder
+                assert '/upload/' in one_file, \
+                    f"Loaded file was resharded not in 'upload' folder on the node {node.name}"
+
+    @timeout_decor(
+        timeout=60,
+        allowed_exceptions=(AssertionError,),
+        message="Waiting for load_and_stream completion message to appear in logs")
+    def validate_load_and_stream_status(self, node, system_log_follower,
```
[the remark this answers is not in the document]
Comment (juliayakovlev) [reply sdcm/utils/common.py:2557]: done

File: unit_tests/test_utils_common.py [hunk 8]
```diff
@@ -84,3 +86,123 @@ def test_update_db_packages_none():
...
+
+    @staticmethod
+    def test_distribute_test_files_to_cluster_nodes():
+        test_cases = [{'nodes': ['node1', 'node2', 'node3'], 'expected_result': 5},
+                      {'nodes': ['node1', 'node2'], 'expected_result': 5},
+                      {'nodes': ['node1', 'node2', 'node3', 'node4'], 'expected_result': 5},
+                      {'nodes': ['node1', 'node2', 'node3', 'node4', 'node5'], 'expected_result': 5},
+                      {'nodes': ['node1', 'node2', 'node3', 'node5', 'node6'], 'expected_result': 5},
+                      ]
+        helpers = LoadUtils()
```
[the remark this answers is not in the document]
Comment (juliayakovlev) [reply unit_tests/test_utils_common.py:193]: removed

File: sdcm/nemesis.py [hunk 9]
```diff
@@ -940,132 +943,67 @@ def disrupt_no_corrupt_repair(self):
...
-            #   /var/lib/scylla/data/keyspace1/standard1-9fbed8d0f8c211ea9bb1000000000000/upload/md-13-big-Data.db:level=0,
-            #   /var/lib/scylla/data/keyspace1/standard1-9fbed8d0f8c211ea9bb1000000000000/upload/md-22-big-Data.db:level=0,
-            #   /var/lib/scylla/data/keyspace1/standard1-9fbed8d0f8c211ea9bb1000000000000/upload/md-15-big-Data.db:level=0,
-            #   /var/lib/scylla/data/keyspace1/standard1-9fbed8d0f8c211ea9bb1000000000000/upload/md-16-big-Data.db:level=0,
-            #   ]. 91MB to 92MB (~100% of original) in 5009ms = 18MB/s. ~370176 total partitions merged to 370150
-            """
-            resharding_logs = list(system_log_follower)
-            assert resharding_logs, "Resharding wasn't run"
-            self.log.debug(f"Found resharding: {resharding_logs}")
+        helpers = LoadUtils()
```
[the remark this answers is not in the document]
Comment (juliayakovlev) [reply sdcm/nemesis.py:961]: removed

File: sdcm/nemesis.py [hunk 10]
```diff
@@ -940,132 +943,67 @@ def disrupt_no_corrupt_repair(self):
...
+                self.log.debug('Key %s does not exist before refresh', key)
             else:
-                self.log.debug('Key 0x32373131364f334f3830 already exists before refresh')
+                self.log.debug('Key %s already exists before refresh', key)
 
             # Executing rolling refresh one by one
             for node in self.cluster.nodes:
-                system_log_follower = do_refresh(node)
-                validate_resharding_after_refresh(system_log_follower)
+                LoadUtils.upload_sstables(node, test_data=test_data)
```
[the remark this answers is not in the document]
Comment (juliayakovlev) [reply sdcm/nemesis.py:1000]: done

File: sdcm/utils/common.py [hunk 11]
```diff
@@ -2420,3 +2421,165 @@ def clear_out_all_exit_hooks():
       teardown.
     """
     threading._threading_atexits.clear()  # pylint: disable=protected-access
+
+
+class LoadUtils:
```
[the remark this answers is not in the document]
Comment (juliayakovlev) [reply sdcm/utils/common.py:2426]: done

File: sdcm/utils/common.py [hunk 12]
```diff
@@ -2420,3 +2421,165 @@ def clear_out_all_exit_hooks():
...
+        # `load_and_stream` parameter is not supported by nodetool yet. This is workaround
+        # https://github.com/scylladb/scylla-tools-java/issues/253
+        load_api_cmd = 'curl -X POST --header "Content-Type: application/json" --header ' \
+                       f'"Accept: application/json" "http://127.0.0.1:10000/storage_service/sstables/{keyspace_name}?' \
+                       f'cf={table_name}&load_and_stream=true"'
+        node.remoter.run(load_api_cmd)
+        return system_log_follower
+
+    @staticmethod
+    def run_refresh(node, test_data: namedtuple) -> List[str]:
```
[the remark this answers is not in the document]
Comment (juliayakovlev) [reply sdcm/utils/common.py:2510]: done

File: sdcm/utils/common.py [hunk 13]
```diff
@@ -2420,3 +2421,165 @@ def clear_out_all_exit_hooks():
...
+    def distribute_test_files_to_cluster_nodes(self, nodes, test_data: list) -> List:
+        """
+        Distribute test sstables over cluster nodes for `load-and-stream` test:
+        the feature allow loading arbitrary sstables that do not belong to a node into the cluster.
+        So we want to distribute sstables randomly.
+        Also add case when the 50% sstables of node2 (from test data cluster) will be loaded on the one node
+        and other 50% - on another
+        """
+        # Load 50% of node2's sstables to the different cluster nodes
+        map_files_to_node = [[test_data.pop(1), self.get_random_item(nodes)],
```
[the remark this answers is not in the document]
Comment (juliayakovlev) [reply sdcm/utils/common.py:2454]: I want this data will be o the different nodes. Fixed

File: sdcm/utils/common.py [hunk 14]
```diff
@@ -2420,3 +2421,165 @@ def clear_out_all_exit_hooks():
...
+                             [test_data.pop(1), self.get_random_item(nodes)]
+                             ]
+
+        invariant, mod = divmod(len(test_data), len(nodes))
```
[the remark this answers is not in the document]
Comment (juliayakovlev) [reply sdcm/utils/common.py:2458]: This case is better to cover in dtest

[the remark this answers is not in the document]
Comment (juliayakovlev) [reply sdcm/utils/common.py:2458]: changed

File: sdcm/utils/sstable_load_utils.py [hunk 15]
```diff
@@ -0,0 +1,162 @@
...
+    def distribute_test_files_to_cluster_nodes(cls, nodes, test_data: list) -> List:
+        """
+        Distribute test sstables over cluster nodes for `load-and-stream` test:
+        the feature allow loading arbitrary sstables that do not belong to a node into the cluster.
+        So we want to distribute sstables randomly.
+        Also add case when the 50% sstables of node2 (from test data cluster) will be loaded on the one node
+        and other 50% - on another
+        """
+        # Load 50% of node2's sstables to the different cluster nodes
+        map_files_to_node = [[test_data.pop(1), nodes[0]], [test_data.pop(1), nodes[1]]]
```
Comment (dkropachev) [inline sdcm/utils/sstable_load_utils.py:41]: test_data can contain from 1 record to 6 records

**Review by dkropachev — CHANGES_REQUESTED** [review]

File: data_dir/test_data_library/load_inventory.py [hunk 16]
```diff
@@ -0,0 +1,116 @@
...
+        return [COLUMN_1_DATA]
+
+    if column_number >= 5:
+        # The snapshot has 5 columns, the snapshot (col=5) can be loaded to table (col > 5).
+        # they rest columns will be filled to 'null'.
+        if load_and_stream:
+            return copy.deepcopy(MULTI_NODE_DATA)
+
+        if big_sstable:
+            return [BIG_SSTABLE_MULTI_COLUMNS_DATA]
```
[the remark this answers is not in the document]
Comment (juliayakovlev) [reply data_dir/test_data_library/load_inventory.py:112]: why?

File: sdcm/utils/sstable_load_utils.py [hunk 17]
```diff
@@ -0,0 +1,162 @@
...
+
+        while test_data:
```
Comment (dkropachev) [inline sdcm/utils/sstable_load_utils.py:43]:
```suggestion
    test_data = test_data.copy()
    nodes_bucket = nodes.copy()
    for data in test_data:
        if not nodes_bucket:
            nodes_bucket = nodes.copy()
         map_files_to_node.append([test_data.pop(), nodes_bucket.pop()])
```

**Review by dkropachev — CHANGES_REQUESTED** [review]

File: sdcm/utils/sstable_load_utils.py (hunk 17 above)
Comment (juliayakovlev) [reply sdcm/utils/sstable_load_utils.py]: done

File: sdcm/utils/sstable_load_utils.py (hunk 15 above)
Comment (juliayakovlev) [reply sdcm/utils/sstable_load_utils.py]: done

File: data_dir/test_data_library/load_inventory.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (juliayakovlev) [reply data_dir/test_data_library/load_inventory.py]: changed

File: data_dir/test_data_library/load_inventory.py [hunk 18]
```diff
@@ -0,0 +1,116 @@
...
+    sstable_md5='9c5dd19cfc78052323995198b0817270',
+    keys_num=501000)]
+
+MULTI_COLUMNS_DATA = TestDataInventory(
+    sstable_url=URL_TEMPLATE % (MULTI_COLUMN_FOLDER, STANDARD1_FILE_NAME),
+    sstable_file=TEMP_FILE_TEMPLATE % STANDARD1_FILE_NAME,
+    sstable_md5='c033a3649a1aec3ba9b81c446c6eecfd',
+    keys_num=1000)
+
+MULTI_NODE_DATA = \
```
[the remark this answers is not in the document]
Comment (juliayakovlev) [reply data_dir/test_data_library/load_inventory.py]: done

File: sdcm/utils/sstable_load_utils.py [hunk 19]
```diff
@@ -0,0 +1,159 @@
...
+
+    @staticmethod
+    def get_random_item(items: list, pop: bool = False) -> Any:
+        if pop:
+            return items.pop(random.randint(0, len(items) - 1))
+
+        return items[random.randint(0, len(items) - 1)]
+
+    @classmethod
+    def distribute_test_files_to_cluster_nodes(cls, nodes, test_data: list) -> List:
```
[the remark this answers is not in the document]
Comment (juliayakovlev) [reply sdcm/utils/sstable_load_utils.py]: done

**Review by dkropachev — APPROVED** [review]

File: sdcm/nemesis.py [hunk 20]
```diff
@@ -940,132 +942,66 @@ def disrupt_no_corrupt_repair(self):
...
-                self.log.debug('Key 0x32373131364f334f3830 already exists before refresh')
+                self.log.debug('Key %s already exists before refresh', key)
 
             # Executing rolling refresh one by one
             for node in self.cluster.nodes:
-                system_log_follower = do_refresh(node)
-                validate_resharding_after_refresh(system_log_follower)
+                SstableLoadUtils.upload_sstables(node, test_data=test_data[0])
+                system_log_follower = SstableLoadUtils.run_refresh(node, test_data=test_data[0])
+                SstableLoadUtils.validate_resharding_after_refresh(system_log_follower, node)
```
Comment (aleksbykov) [inline sdcm/nemesis.py:1000]: just small notice, 
you have method:   SstableLoadUtils.validate_load_and_stream_status(load_on_node, system_log_follower) where first param is node , second is log_follower

may by it is better to have same parameter order for this method too?

Comment (juliayakovlev) [reply sdcm/nemesis.py:1000]: > just small notice,
> you have method: SstableLoadUtils.validate_load_and_stream_status(load_on_node, system_log_follower) where first param is node , second is log_follower
> 
> may by it is better to have same parameter order for this method too?

@aleksbykov I think you confised between 2 methods: **validate_resharding_after_refresh** and **validate_load_and_stream_status**. It's different methods

Comment (aleksbykov) [reply sdcm/nemesis.py:1000]: the method are different, but i think that order of parameters in method signature could be the same.

Comment (juliayakovlev) [reply sdcm/nemesis.py]: done

**Review by bentsi — APPROVED** [review]

**Pull request merged**